module/persistent.lua

```diff
@@ -0,0 +1,290 @@
+                tag.volatile = self.restored_settings.tags[index].volatile
+                tag.gap = self.restored_settings.tags[index].gap
+                tag.gap_single_client = self.restored_settings.tags[index].gap_single_client
+                tag.master_fill_policy = self.restored_settings.tags[index].master_fill_policy
+                tag.master_count = self.restored_settings.tags[index].master_count
+                tag.column_count = self.restored_settings.tags[index].column_count
+                if self.restored_settings.tags[index].selected == true then
+                    awful.tag.viewtoggle(tag)
+                    selected_tag = true
+                end
+            end
+        end
+    end
+
+    if selected_tag == false then
+        -- awful.tag.viewtoggle(capi.root.tags()[1])
+    end
+end
+
+local function restore_clients(self, args)
+    args.create_clients = args.create_clients ~= nil and args.create_clients or false
+
+    if args.create_clients == true then
+        for pid, client in pairs(self.restored_settings.clients) do
+            awful.spawn.easy_async_with_shell(string.format("ps aux | grep '%s' | grep -v 'grep'", client.command), function(stdout)
+                if stdout == "" or stdout == nil then
+                    local new_pid = awful.spawn(client.command, false)
+                    self.restored_settings.clients[tostring(new_pid)] = self.restored_settings.clients[pid]
+                    self.restored_settings.clients[pid] = nil
+                end
+            end)
+        end
+
+        gtimer {
+            timeout = 3,
+            call_now = false,
+            autostart = true,
+            single_shot = true,
+            callback = function()
+                reapply_clients(self)
+            end
+        }
+    else
+        reapply_clients(self)
+    end
+end
+
+function persistent:save(args)
+    args = args or {}
+
+    args.save_tags = args.save_tags == nil and true or args.save_tags
+    args.save_clients = args.save_clients == nil and true or args.save_clients
+
+    if args.save_tags == true then
+        save_tags(self)
+    end
+    if args.save_clients ==  true then
+        save_clients(self)
+    end
+
+    local cache = gfilesystem.get_xdg_cache_home()
+    if not gfilesystem.dir_readable(cache .. "awesome") then
+        gfilesystem.make_directories(cache .. "awesome")
+    end
+
+    local json_settings = json.encode(self.settings, { indent = true })
+    local path = cache .. "awesome/persistent.json"
+    awful.spawn.with_shell("echo '" .. json_settings .. "'" .. " > " .. path)
+end
+
+function persistent:restore(args)
+    args = args or {}
+
+    args.restore_tags = args.restore_tags == nil and true or args.restore_tags
+    args.restore_clients = args.restore_clients == nil and true or args.restore_clients
+
+    local path = gfilesystem.get_xdg_cache_home() .. "awesome/persistent.json"
+    awful.spawn.easy_async_with_shell("cat " .. path, function(stdout)
```

Comment:
That looks really weird tbh. Is there a reason why you can't read the file with luas `io` library?